libdrgn/arch_s390x.c

```diff
@@ -237,6 +237,223 @@ linux_kernel_get_initial_registers_s390x(const struct drgn_object *task_obj,
+	{ .bits = 11, .shift = 42, .entries = 2048 },	/* RSTE */
+	{ .bits = 11, .shift = 53, .entries = 2048 },	/* RFTE */
+};
+
+#define REGION_INVALID		0x20
+#define SEGMENT_INVALID		0x20
+#define PAGE_INVALID		0x400
+
+#define REGION_LARGE		0x400
+#define SEGMENT_LARGE		0x400
+#define PAGE_LARGE		0x800
+
+static struct drgn_error *
+linux_kernel_pgtable_iterator_create_s390(struct drgn_program *prog,
+					  struct pgtable_iterator **ret)
+{
+	struct pgtable_iterator_s390x *it = malloc(sizeof(*it));
+	if (!it)
+		return &drgn_enomem;
+
+	*ret = &it->it;
+	return NULL;
+}
+
+static void
+linux_kernel_pgtable_iterator_destroy_s390(struct pgtable_iterator *_it)
+{
+	free(container_of(_it, struct pgtable_iterator_s390x, it));
+}
+
+static void
+linux_kernel_pgtable_iterator_init_s390(struct drgn_program *prog,
+					struct pgtable_iterator *_it)
+{
+	struct pgtable_iterator_s390x *it =
+		container_of(_it, struct pgtable_iterator_s390x, it);
+	for (int i = 0; i < 5; i++)
+		it->pagetable[i].addr = -1UL;
+	it->levels = 3;
+}
+
+static int
+entry_is_invalid(uint64_t entry, int level)
+{
+	switch (level) {
+	case 1:
+		return entry & PAGE_INVALID;
+	case 2:
+		return entry & SEGMENT_INVALID;
+	case 3:
+	case 4:
+	case 5:
+		return entry & REGION_INVALID;
+	default:
+		break;
+	}
+	return 1;
+}
+
+static int
+get_mask(struct pgtable_iterator_s390x *it, int level)
+{
+	return (1 << dat_levels[level].bits) - 1;
+}
+
+static int
+get_index(struct pgtable_iterator_s390x *it, int level, uint64_t va)
+{
+	return (va >> dat_levels[level].shift) & get_mask(it, level);
+}
+
+static int
+get_table_length(uint64_t entry, int level)
+{
+	switch(level) {
+	case 3:
+	case 4:
+	case 5:
+		return ((entry & 3) + 1) << 9;
+	default:
+		return 2048;
+	}
+}
+
+static int
+get_table_offset(uint64_t entry, int level)
+{
+	switch(level) {
+	case 3:
+	case 4:
+	case 5:
+		return ((entry >> 6) & 3) << 9;
+	default:
+		return 0;
+	}
+}
+
+static uint64_t
+get_table_address(uint64_t entry, int level, bool *is_large)
+{
+	*is_large = false;
+	switch (level) {
+	case 2: /* STE */
+		if (entry & SEGMENT_LARGE) {
+			*is_large = true;
+			return entry & ~((1UL << 44) - 1);
+		} else {
+			return entry & ~0x7ffUL;
+		}
+	case 3: /* RTTE */
+		if (entry & REGION_LARGE) {
+			*is_large = true;
+			return entry & ~((1UL << 31) - 1);
+		} else {
+			return entry & ~0xfffUL;
+		}
+	default:
+		return entry & ~0xfffUL;
+	}
+}
+
+static struct drgn_error *
+linux_kernel_pgtable_iterator_next_s390(struct drgn_program *prog,
+					struct pgtable_iterator *_it,
+					uint64_t *virt_addr_ret,
+					uint64_t *phys_addr_ret)
+{
+	struct pgtable_iterator_s390x *it =
+		container_of(_it, struct pgtable_iterator_s390x, it);
+	const uint64_t va = it->it.virt_addr;
+	uint64_t table = _it->pgtable & ~0xfffUL;
+	int levels = 2, length = 2048, offset = 0;
+	uint64_t entry;
+
+	struct drgn_error *err = drgn_program_read_memory(prog, &entry,
+							  table, 8, true);
```

Comment:
This doesn't byte swap the value if the host is little-endian. `drgn_program_read_u64()` will do that for you. Also, the initial page table is a virtual address, so `physical` should be `false` (but from what I understand, the addresses are equal on s390x anyways, so this would be more for documentation/consistency).

```suggestion
	struct drgn_error *err = drgn_program_read_u64(prog, table, 8, false,
						       &entry);
```

Finally, could you please add a comment here explaining what we are doing here? I.e., getting the ASCE bits in a Linux kernel-specific way.